gtda/homology/_utils.py

```diff
@@ -1,52 +1,64 @@
-""" Utilities function for persistent homology."""
+"""Utility functions for persistent homology."""
 # License: GNU AGPLv3
 
 import numpy as np
-from joblib import Parallel, delayed
 
 
-def _pad_diagram(Xd, homology_dimensions, max_n_points, min_values):
-    for dim in homology_dimensions:
-        n_points = len(Xd[dim])
-        n_points_to_pad = max_n_points[dim] - n_points
-        if n_points == 0 and n_points_to_pad == 0:
-            n_points_to_pad = 1
-
-        if n_points_to_pad > 0:
-            padding = ((0, n_points_to_pad), (0, 0))
-            Xd[dim] = np.pad(Xd[dim], padding, 'constant')
-            Xd[dim][-n_points_to_pad:, :] = \
-                [min_values[dim], min_values[dim]]
-
-    # Add dimension as the third elements of each (b, d) tuple
-    Xd = [np.insert(Xd[dim], 2, dim, axis=1)
-          for dim in homology_dimensions]
-
-    Xd = np.vstack(Xd)
-    return Xd
+def _postprocess_diagrams(
+        Xt, format, homology_dimensions, infinity_values
+        ):
+    # NOTE: `homology_dimensions` must be sorted in ascending order
+    def replace_infinity_values(subdiagram):
+        np.nan_to_num(subdiagram, posinf=infinity_values, copy=False)
+        return subdiagram[subdiagram[:, 0] < subdiagram[:, 1]]
 
+    # Replace np.inf with infinity_values and turn into list of dictionaries
+    # whose keys are the dimensions
+    if format in ["ripser", "flagser"]:  # Input is list of list of subdiagrams
+        # In H0, remove one infinite bar placed at the end by ripser
+        slices = {dim: slice(None) if dim else slice(None, -1)
+                  for dim in homology_dimensions}
+        Xt = [{dim: replace_infinity_values(diagram[dim][slices[dim]])
+               for dim in homology_dimensions} for diagram in Xt]
+    elif format == "gudhi":  # Input is list of list of [dim, (birth, death)]
+        # In H0, remove one infinite bar placed at the beginning by GUDHI
```

Comment:
Is there always an infinte bar here? I am asking because I can't remember why I did not do this.

Reply:
I will have a deeper look! (Maybe Adelie knows?)

Reply:
What happens if you work with binary images and you try to calculate cubical homology of an interely `False` image? Do you have a bug in the next line?

Reply:
I think you'll get an empty array and no error, but I should test.

Reply:
I just tested this:
```
import numpy as np
from gtda.homology import CubicalPersistence
images = np.full((1, 20, 20), fill_value=False)
CubicalPersistence().fit_transform(images)
```
and the output is
```
array([[[0., 0., 0.],
        [0., 0., 1.]]])
```
which seems fine to me if we remove the infinite bar.

Reply:
I just tested this:
```
import numpy as np
from gtda.homology import CubicalPersistence
images = np.full((1, 20, 20), fill_value=False)
CubicalPersistence().fit_transform(images)
```
and the output is
```
array([[[0., 0., 0.],
        [0., 0., 1.]]])
```
which seems fine to me if we remove the infinite bar.